### PR TITLE
Parse message calls as filters

### PR DESCRIPTION
**Pull request by micgro42**
`var|filter` is not a thing anymore. This now parses the new syntax of [Vue 3: Turn filters into methods (I28310c09)](https://gerrit.wikimedia.org/r/c/mediawiki/extensions/WikibaseLexeme/+/843986) with minimal changes to the interface.

TODO:
* meaningful commit message
* delete now dead code?
* do we want to do bigger refactorings?

Bug: [⚓ T321595 Migrate filters in WikibaseLexeme Vue templates to something Vue 3 compatible](https://phabricator.wikimedia.org/T321595)

### CODE REVIEW

**Comment by micgro42** [comment]
> Hm, is method syntax really the way we want to go? I thought the directive syntax seemed more promising (after you brought it up).

On the one hand yes, on the other hand it will require additional work for directionality and language name (though I guess we could still go with the method syntax for those cases).

I wonder if it would be possible to go with the minimally invasive version now, and reconsider things more fundamentally when we can switch to proper SSR.

**Comment by lucaswerkmeister** [comment]
Well, I didn’t think that adding support for method syntax would be the minimally invasive approach ^^

I initially thought I’d prefer a clearer separation between filters and methods, but maybe it’s actually good to start with a version of the library that supports both of them in the same way, so that we can easily transition from filters to methods in WikibaseLexeme. But I think this should at least mention the filter/method overlap in more places, including:

- rename `$filters` to something like `$filtersAndMethods` in at least `Templating`, the public entry point (maybe also other classes)
- add phpdoc to `FilterCall` mentioning that it also includes method calls
- probably also mention it in the README?

I think this should also be a semver-breaking release, i.e. version 2.0.0. (We can require `^1.0.2 || ^2.0.0` in WikibaseLexeme.) If we later publish a version where we’ve removed the filter support, that could be version 3.0.0.

**Comment by micgro42** [comment]
> Well, I didn’t think that adding support for method syntax would be the minimally invasive approach ^^

I'm very open to alternatives that require only little changes to this library :).


> I initially thought I’d prefer a clearer separation between filters and methods, but maybe it’s actually good to start with a version of the library that supports both of them in the same way, so that we can easily transition from filters to methods in WikibaseLexeme. But I think this should at least mention the filter/method overlap in more places, including:
> 
>     * rename `$filters` to something like `$filtersAndMethods` in at least `Templating`, the public entry point (maybe also other classes)
> 
>     * add phpdoc to `FilterCall` mentioning that it also includes method calls
> 
>     * probably also mention it in the README?
> 
> I think this should also be a semver-breaking release, i.e. version 2.0.0. (We can require `^1.0.2 || ^2.0.0` in WikibaseLexeme.) If we later publish a version where we’ve removed the filter support, that could be version 3.0.0.

Happy to do all the documentation updates and also rename the parameters. That all sounds sensible to me. Though that would not _yet_ be breaking semver right? All code we had so far should continue working, we're just parsing now more syntax that would previously have thrown an error.
That being said, in this case we probably don't want to _change_ the widget fixtures, but duplicate them so that we have both "legacy" fixtures and the now normal Vue3 ones.

The later version would definitively be a breaking change, though.

**Comment by lucaswerkmeister** [comment]
Hm, conceptually this feels like a major change to me, because we’re now treating methods and filters as equivalent (which they aren’t in Vue, though I only noticed that partway through my WIP change in WikibaseLexeme ^^) – but you have a point that we didn’t support methods at all before. So maybe a minor version bump is enough.

Good point about copying instead of changing the fixtures.

File: tests/php/FilterExpressionParsing/FilterParserTest.php [hunk 1]
```diff
@@ -69,6 +69,10 @@ public function provideParseCases() {
 				'{prop1:var1, prop2:var2}',
 				new ParseResult( [ '{prop1:var1, prop2:var2}' ], [] )
 			],
+			'method call with single argument' => [
+				'filter(var1)',
+				new ParseResult( [ 'var1' ], [ new FilterCall( 'filter', [] ) ] )
+			],
 			'variable with filter having no arguments' => [
 				'var1|filter',
 				new ParseResult( [ 'var1' ], [ new FilterCall( 'filter', [] ) ] )
```
Comment (lucaswerkmeister) [inline tests/php/FilterExpressionParsing/FilterParserTest.php:78]: It’s weird how, just below this test, you can see that we actually have support for parenthesized calls already (`var1|filter(var2)`). (I had no idea Vue supported this, and we certainly don’t seem to use it in WikibaseLexeme – so much for only covering a subset of the Vue syntax that’s used by WikibaseLexeme…) But this seems to rely on the parser’s `$this->nowParsingFilterList` state: all parentheses before the first filter pipe are “grouping” parentheses, and all parentheses after the first filter pipe are “call” parentheses. I don’t think we can easily repurpose this code for parsing method calls, at least not while keeping support for filter syntax as well, so I think the regex hack is good enough for now.

Once we drop filter support entirely, I think we can remove the whole `FilterExpressionParsing` step, and instead move the regex into the `BasicJsExpressionParser`? (Which is quite basic indeed – lots of the syntax that the filter expression parser supports and tests, like arrays and objects, don’t actually work in the JS part as far as I can tell.)

Comment (micgro42) [reply tests/php/FilterExpressionParsing/FilterParserTest.php:78]: Yes, I noticed the existing syntax as well. But making use of that without breaking anything seemed like the more intricate challenge

**Comment by micgro42** [comment]
@lucaswerkmeister This should now be ready for review. Not sure how elaborate we want to be with the documentation. Let me know if you expected it to be more elaborate in some places.

File: tests/integration/fixture/gloss_widget.html [hunk 2]
```diff
@@ -26,8 +26,8 @@
             </table>
         </div>
         <div class="wikibase-lexeme-sense-glosses-controls">
-            <button type="button" class="wikibase-lexeme-sense-glosses-control" v-if="!inEditMode" v-on:click="edit" :disabled="isSaving">{{message('wikibase-edit')}}</button>
-            <button type="button" class="wikibase-lexeme-sense-glosses-control" v-if="inEditMode" v-on:click="save" :disabled="isSaving">{{message('wikibase-save')}}</button>
+            <button type="button" class="wikibase-lexeme-sense-glosses-control" v-if="!inEditMode" v-on:click="edit" :disabled="isSaving">{{ message('wikibase-edit') }}</button>
+            <button type="button" class="wikibase-lexeme-sense-glosses-control" v-if="inEditMode" v-on:click="save" :disabled="isSaving">{{message( 'wikibase-save' )}}</button>
```
Comment (lucaswerkmeister) [inline tests/integration/fixture/gloss_widget.html:30]: I don’t think we should test the spacing here – we probably won’t use it like that (at least not inconsistently) in WikibaseLexeme itself, and I’d prefer to limit differences between these fixtures and the real WikibaseLexeme templates. (Though they’ve apparently already diverged quite a bit 😕) The unit test below should be enough IMHO.

**Review by lucaswerkmeister — COMMENTED** [review]
Looks pretty good and I think the documentation is sufficient. I’ll try it out with WikibaseLexeme later.

File: tests/integration/fixture/gloss_widget.html (hunk 2 above)
Comment (micgro42) [reply tests/integration/fixture/gloss_widget.html]: Done.

**Review by lucaswerkmeister — APPROVED** [review]
Tried it locally, seems to work \o/

**Comment by micgro42** [comment]
Will merge after Travis is happy :+1:

**Pull request merged**